File: js/src/infrastructure/repositories/InMemoryIdempotencyRepository.ts
import { TransferMoneyResponse } from "../../application/dtos/TransferMoneyCommand";
import { IIdempotencyRepository } from "../../application/repositories/IIdempotencyRepository";

export class InMemoryIdempotencyRepository implements IIdempotencyRepository {
  private cache = new Map<string, TransferMoneyResponse>();

  public async find(key: string): Promise<TransferMoneyResponse | null> {
    const cached = this.cache.get(key);
    return cached ? { ...cached } : null;
  }

  public async save(key: string, response: TransferMoneyResponse): Promise<void> {
    this.cache.set(key, { ...response });
  }

  public clear(): void {
    this.cache.clear();
  }
}

File: py/src/infrastructure/repositories/in_memory_idempotency_repository.py
from src.application.dtos.transfer_dto import TransferMoneyResponse
from src.application.repositories.idempotency_repository import IdempotencyRepository

class InMemoryIdempotencyRepository(IdempotencyRepository):
    def __init__(self):
        self._cache: dict[str, TransferMoneyResponse] = {}

    def find(self, key: str) -> TransferMoneyResponse | None:
        response = self._cache.get(key)
        if response:
            # Return a copy to mimic DB retrieval isolation
            return TransferMoneyResponse(
                success=response.success,
                transaction_id=response.transaction_id,
                error=response.error,
                balance_after_transfer=response.balance_after_transfer
            )
        return None

    def save(self, key: str, response: TransferMoneyResponse) -> None:
        self._cache[key] = response

    def clear(self) -> None:
        self._cache.clear()
